hacep/src/main/java/org/kie/hacep/core/KieSessionContext.java

```diff
@@ -15,31 +15,49 @@
  */
 package org.kie.hacep.core;
 
+import java.util.concurrent.TimeUnit;
+
 import org.kie.api.runtime.KieSession;
+import org.kie.api.time.SessionClock;
+import org.kie.api.time.SessionPseudoClock;
 import org.kie.hacep.consumer.FactHandlesManager;
 import org.kie.hacep.core.infra.SnapshotInfos;
 
 public class KieSessionContext {
 
     private KieSession kieSession;
 
+    private SessionPseudoClock clock;
+
     private FactHandlesManager fhManager;
 
     public KieSession getKieSession() {
         return kieSession;
     }
 
     public void initFromSnapshot(SnapshotInfos infos) {
-        this.kieSession = infos.getKieSession();
+        setKieSession(infos.getKieSession());
         this.fhManager = infos.getFhManager();
     }
 
     public void init(KieSession newKiession) {
-        this.kieSession = newKiession;
+        setKieSession(newKiession);
         this.fhManager = new FactHandlesManager(newKiession);
     }
 
+    private void setKieSession(KieSession kieSession) {
+        this.kieSession = kieSession;
+        SessionClock clock = kieSession.getSessionClock();
+        if (clock instanceof SessionPseudoClock) {
+            this.clock = (SessionPseudoClock) clock;
+        }
+    }
+
     public FactHandlesManager getFhManager() {
         return fhManager;
     }
+
+    public void setClockAt(long time) {
+        clock.advanceTime(time - clock.getCurrentTime(), TimeUnit.MILLISECONDS);
```

Comment:
This can throw NPE if not pseudo (line 51)

Reply:
If this NPEs it means that we're doing CEP without a pseudo-clock (the method isn't called if the fact is not an event) and in this case I'd prefer to fail fast with a NPE instead of doing defensive programming.

Reply:
I agree to fail fast but I prefer not to have a NPE but a more meaningful message :)